From: Stained Glass OS <ke7oxh@gmail.com>
uiautomationcore: CreatePropertyConditionEx and ElementFromPoint

CreatePropertyConditionEx passes the IgnoreCase flag on (other flags are
E_INVALIDARG); ElementFromPoint(BuildCache) returns the element of the
window under the point.
---
--- a/dlls/uiautomationcore/uia_com_client.c
+++ b/dlls/uiautomationcore/uia_com_client.c
@@ -5219,8 +5219,22 @@ static HRESULT WINAPI uia_iface_ElementFromHandle(IUIAutomation6 *iface, UIA_HWN
 
 static HRESULT WINAPI uia_iface_ElementFromPoint(IUIAutomation6 *iface, POINT pt, IUIAutomationElement **out_elem)
 {
-    FIXME("%p, %s, %p: stub\n", iface, wine_dbgstr_point(&pt), out_elem);
-    return E_NOTIMPL;
+    HWND hwnd;
+
+    TRACE("%p, %s, %p\n", iface, wine_dbgstr_point(&pt), out_elem);
+
+    if (!out_elem)
+        return E_POINTER;
+    *out_elem = NULL;
+    /* The deepest window under the point; the provider tree inside a window is not searched. */
+#ifdef SG_MUTANT_UIA_POINT_ROOT
+    hwnd = GetDesktopWindow();
+#else
+    hwnd = WindowFromPoint(pt);
+#endif
+    if (!hwnd)
+        return UIA_E_ELEMENTNOTAVAILABLE;
+    return IUIAutomation6_ElementFromHandle(iface, (UIA_HWND)hwnd, out_elem);
 }
 
 static HRESULT uia_get_focused_element(IUIAutomation6 *iface, IUIAutomationCacheRequest *cache_req,
@@ -5324,8 +5338,17 @@ static HRESULT WINAPI uia_iface_ElementFromHandleBuildCache(IUIAutomation6 *ifac
 static HRESULT WINAPI uia_iface_ElementFromPointBuildCache(IUIAutomation6 *iface, POINT pt,
         IUIAutomationCacheRequest *cache_req, IUIAutomationElement **out_elem)
 {
-    FIXME("%p, %s, %p, %p: stub\n", iface, wine_dbgstr_point(&pt), cache_req, out_elem);
-    return E_NOTIMPL;
+    IUIAutomationElement *elem;
+    HRESULT hr;
+
+    TRACE("%p, %s, %p, %p\n", iface, wine_dbgstr_point(&pt), cache_req, out_elem);
+
+    if (!out_elem)
+        return E_POINTER;
+    *out_elem = NULL;
+    if (FAILED(hr = IUIAutomation6_ElementFromPoint(iface, pt, &elem)))
+        return hr;
+    return element_with_cache(elem, cache_req, out_elem);
 }
 
 static HRESULT WINAPI uia_iface_GetFocusedElementBuildCache(IUIAutomation6 *iface,
@@ -5485,8 +5508,16 @@ static HRESULT WINAPI uia_iface_CreatePropertyCondition(IUIAutomation6 *iface, P
 static HRESULT WINAPI uia_iface_CreatePropertyConditionEx(IUIAutomation6 *iface, PROPERTYID prop_id, VARIANT val,
         enum PropertyConditionFlags flags, IUIAutomationCondition **out_condition)
 {
-    FIXME("%p, %d, %s, %#x, %p: stub\n", iface, prop_id, debugstr_variant(&val), flags, out_condition);
-    return E_NOTIMPL;
+    TRACE("%p, %d, %s, %#x, %p\n", iface, prop_id, debugstr_variant(&val), flags, out_condition);
+
+    if (!out_condition)
+        return E_POINTER;
+    *out_condition = NULL;
+#ifndef SG_MUTANT_UIA_FLAGS_UNCHECKED
+    if (flags & ~PropertyConditionFlags_IgnoreCase)
+        return E_INVALIDARG;
+#endif
+    return create_uia_property_condition_iface(out_condition, prop_id, val, flags);
 }
 
 /* Stained Glass OS: And conditions (Opera's installer: class name and
